### ISSUE TITLE
Loading a new project in the same window doesn't terminate Rust Analyzer instance attached to the previous project

### ISSUE THREAD

**Issue by antoinecharbonneau** [zed-industries/zed#17480]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Tested on Fedora with Rust Analyzer, might apply more broadly.

Loading different projects one after the other within a single window of Zed doesn't kill the previous Rust analyzer instance. They end up accumulating and running your system out of memory. The dangling rust-analyzer instances don't use up CPU, but don't free the memory as they should. This problem eventually leads to the system freezing from memory starvation as rust-analyzer quickly gobbles up 1 or 2 Gb of memory per instance.

To reproduce the bug quickly, have two projects to quickly toggle in the recent project list.
Make sure rust-analyzer starts loading each time by opening a rust file.

It might be useful to be able to have LSPs reused when reopening a project that was opened during the session, but Zed should make sure it reuses it instead of accumulating instances. Alternatively, the previous LSP should be terminated on project 

Logs included had the application opened on a project, with 4 project switch. Logs are too long for this, I can include them if deemed necessary, though I haven't found anything of interest from skimming through them (not an expert on Zed though).

```
$ ps -aux | grep rust-analyzer

user     6226  3.7  6.4 2298184 1040828 ?     Sl   00:22   0:40 /home/user/.local/share/zed/languages/rust-analyzer/rust-analyzer-2024-09-02
user     6297  0.1  0.4 371080 69876 ?        S    00:22   0:01 /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/libexec/rust-analyzer-proc-macro-srv
user    16664  6.4  7.4 2428916 1208164 ?     Sl   00:29   0:39 /home/user/.local/share/zed/languages/rust-analyzer/rust-analyzer-2024-09-02
user    16707  0.2  0.4 371280 74372 ?        S    00:29   0:01 /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/libexec/rust-analyzer-proc-macro-srv
user    18548  5.7  6.6 2310132 1074956 ?     Sl   00:30   0:33 /home/user/.local/share/zed/languages/rust-analyzer/rust-analyzer-2024-09-02
user    18681  0.1  0.4 371484 70292 ?        S    00:30   0:01 /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/libexec/rust-analyzer-proc-macro-srv
user    19772  7.4  7.5 2419700 1220096 ?     Sl   00:30   0:41 /home/user/.local/share/zed/languages/rust-analyzer/rust-analyzer-2024-09-02
user    19890  0.2  0.4 371020 73812 ?        S    00:30   0:01 /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/libexec/rust-analyzer-proc-macro-srv
user    22115  7.5  6.4 2294076 1038188 ?     Sl   00:31   0:39 /home/user/.local/share/zed/languages/rust-analyzer/rust-analyzer-2024-09-02
user    22241  0.2  0.4 371148 69732 ?        S    00:31   0:01 /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/libexec/rust-analyzer-proc-macro-srv
```

### Environment

Zed: v0.151.1 (Zed)
OS: Linux Wayland fedora 40
Memory: 15.4 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics 620 (KBL GT2) || Intel open-source Mesa driver || Mesa 24.1.5

**Comment by antoinecharbonneau** [comment]
Whenever someone looks into this issue, if they would be so kind as to walk me through their fixing process as I have been circling around the source code not knowing how and where exactly the language server should be stopped. My quest has ended in `workspace/src/workspace.rs:1135`

Thanks

**Comment by antoinecharbonneau** [comment]
From my testing, @mrnugget fixed it in [54dd408](https://github.com/zed-industries/zed/commit/54dd40878fda24701d98ac592af4581186467bf7).

**Comment by mrnugget** [comment]
> From my testing, @mrnugget fixed it in [54dd408](https://github.com/zed-industries/zed/commit/54dd40878fda24701d98ac592af4581186467bf7).

Really? That was just a precursor to fixing this issue. 

From my understanding, this code would only trigger when the app is shut down:

https://github.com/zed-industries/zed/blob/894866da9478003e77458e7a849353312d4c282c/crates/project/src/lsp_store.rs#L257

I'm working on triggering it when the project is dropped. But maaaybe there's already some logic that drops the LSP Store when the workspace is dropped and thus the processes get terminated? I still think we should change that logic here to terminate the servers when the store is released.

**Comment by mrnugget** [comment]
Figured it out. Yeah, it's RAII. We do shutdown the language server when the reference is dropped:

https://github.com/zed-industries/zed/blob/894866da9478003e77458e7a849353312d4c282c/crates/lsp/src/lsp.rs#L1079-L1088
